### ISSUE TITLE
Ignition Reads Config From URL Before NIC Driver is Loaded

### ISSUE THREAD

**Issue by danehans** [coreos/bugs#1618]
# Issue Report #

### CoreOS Version ###

```
1192.0.0
```

### Environment ###

Physical Cisco UCS servers being managed by coreos-baremetal (bootcfg). Instead of using the onboard Intel NIC's, I am using  a 2-port Cisco [VIC 1277 card](http://www.cisco.com/c/en/us/products/collateral/interfaces-modules/unified-computing-system-adapters/datasheet-c78-732515.html) that presents virtual interfaces to the operating system.

### Expected Behavior ###

NIC driver loads then Ignition makes a call to the http config url being served by bootcfg.

### Actual Behavior ###

Ignition calls the http config url before the NIC drivers are loaded. This causes the url call to fail b/c networking is not established.

### Reproduction Steps ###

  1. Setup a bootcfg server to serve ignition templates to managed servers. Use the coreos dnsmasq Docker image to provide dhcp/bootp.
  2. Use Cisco UCS servers with Cisco VIC cards (instead of onboard Intel NIC) for managed servers.
  3. Use port-0 of VIC for management. This mac should be used by bootcfg for group matching.
  3. Boot managed servers. Observe bootcfg provision CoreOS to the managed server. After CoreOS is provisioned, the Ignition template should be used to serve the config to the managed server.
  4. The managed server is unable to hit the config url of bootcfg b/c the NIC drivers are loaded after Ignition's config call. The call fails and CoreOS boots into emergency mode.

### Other Information ###

Screen shots:
![screen shot 2016-10-17 at 3 48 33 pm](https://cloud.githubusercontent.com/assets/1878195/19461026/d6c3f6ac-9495-11e6-9aa7-bd42f6b5bc61.png)
![screen shot 2016-10-17 at 3 49 04 pm](https://cloud.githubusercontent.com/assets/1878195/19461028/e0cf2112-9495-11e6-854c-92df9d940bdf.png)
![screen shot 2016-10-17 at 3 50 00 pm](https://cloud.githubusercontent.com/assets/1878195/19461034/e5a2165e-9495-11e6-9da0-f4756084b44d.png)

**Comment by crawford** [comment]
Ignition won't wait for the network to come up before it starts. Instead, it will retry HTTP[S] operations when they fail. In this case, I can't quite tell what's going on since the logs are limited.

  - Does the network ever come up?
  - Is 1192.0.0 the version being PXE-booted by bootcfg or the target version to be installed?
  - Is there anyway you can get the full Ignition logs (`journalctl -t ignition --no-pager`)?

/cc @dghubble

**Comment by danehans** [comment]
@crawford thanks for the explanation.

- The network does come up. I am able to ping the address of the target server from a remote host.
- I tried 1053.2.0, 1081.5.0 and 1192.0.0 on the target servers and I observe the same behavior.
- Let me try to get the full Ignition logs. I'm a bit limited with the remote console.

I made changes to the lab setup to work around the issue, so give me time to revert the configurations.

**Comment by crawford** [comment]
Okay, if the network comes up and this is failing, I suspect Ignition just isn't waiting long enough. We'll know for sure from the logs.

**Comment by danehans** [comment]
@dghubble @crawford thanks for your patience while i recreated the issue. It appears the target server hits the bootcfg server on it's 5th attempt. However, the config is not pulled b/c it can;t be parsed:
```
$ systemctl status bootcfg -l
● bootcfg.service - coreos-baremetal project
   Loaded: loaded (/etc/systemd/system/bootcfg.service; enabled; vendor preset: disabled)
   Active: active (running) since Tue 2016-10-18 19:39:39 GMT; 23h ago
 Main PID: 19223 (docker)
   Memory: 5.2M
   CGroup: /system.slice/bootcfg.service
           └─19223 /usr/bin/docker run --net=host --rm -v /home/localadmin/code/go/src/github.com/coreos/coreos-baremetal/examples:/var/lib/bootcfg:Z -v /home/localadmin/code/go/src/github.com/coreos/coreos-baremetal/examples/groups/etcd:/var/lib/bootcfg/groups:Z -v /home/localadmin/code/go/src/github.com/coreos/coreos-baremetal/examples/etc/bootcfg:/etc/bootcfg:Z quay.io/coreos/bootcfg:latest -address=10.30.118.133:8080 -log-level=debug -ca-file=/etc/bootcfg/ca.crt -cert-file=/etc/bootcfg/server.crt -key-file=/etc/bootcfg/server.key -rpc-address=10.30.118.133:8081

Oct 19 18:49:13 gem-master01 docker[19223]: 2016-10-19 18:49:13.889469 D | http: HTTP GET /ignition?uuid=4cefeb51-7cbf-4d4c-827c-784e332b6071&mac=00-3a-7d-cf-db-bd
Oct 19 18:49:13 gem-master01 docker[19223]: 2016-10-19 18:49:13.898577 E | http: error parsing Ignition config: path not absolute
Oct 19 18:58:32 gem-master01 docker[19223]: 2016-10-19 18:58:32.982181 D | http: HTTP GET /boot.ipxe
Oct 19 18:58:32 gem-master01 docker[19223]: 2016-10-19 18:58:32.995926 D | http: HTTP GET /ipxe?uuid=4cefeb51-7cbf-4d4c-827c-784e332b6071&mac=00-3a-7d-cf-db-bd&domain=&hostname=&serial=FCH2014V0DQ
Oct 19 18:58:33 gem-master01 docker[19223]: 2016-10-19 18:58:33.002390 D | http: HTTP GET /assets/coreos/1192.0.0/coreos_production_pxe.vmlinuz
Oct 19 18:59:06 gem-master01 docker[19223]: 2016-10-19 18:59:06.539405 D | http: HTTP GET /assets/coreos/1192.0.0/coreos_production_pxe_image.cpio.gz
Oct 19 19:04:02 gem-master01 docker[19223]: 2016-10-19 19:04:02.244073 D | http: HTTP GET /ignition?uuid=4cefeb51-7cbf-4d4c-827c-784e332b6071&mac=00-3a-7d-cf-db-bd
Oct 19 19:04:02 gem-master01 docker[19223]: 2016-10-19 19:04:02.254914 E | http: error parsing Ignition config: path not absolute
Oct 19 19:04:02 gem-master01 docker[19223]: 2016-10-19 19:04:02.323770 D | http: HTTP GET /ignition?uuid=4cefeb51-7cbf-4d4c-827c-784e332b6071&mac=00-3a-7d-cf-db-bd
Oct 19 19:04:02 gem-master01 docker[19223]: 2016-10-19 19:04:02.332884 E | http: error parsing Ignition config: path not absolute
```
I am using [this](https://github.com/coreos/coreos-baremetal/blob/master/examples/ignition/k8s-controller.yaml) Ignition config.

Here is my bootcfg profile for the target server:
```
$ cat examples/profiles/k8s-controller.json 
{
  "id": "k8s-controller",
  "name": "Kubernetes Controller",
  "boot": {
    "kernel": "/assets/coreos/1192.0.0/coreos_production_pxe.vmlinuz",
    "initrd": ["/assets/coreos/1192.0.0/coreos_production_pxe_image.cpio.gz"],
    "cmdline": {
      "root": "/dev/sda1",
      "coreos.config.url": "http://10.30.118.133:8080/ignition?uuid=${uuid}&mac=${net0/mac:hexhyp}",
      "coreos.autologin": "",
      "coreos.first_boot": "1"
    }
  },
  "cloud_id": "",
  "ignition_id": "k8s-controller.yaml"
}
```

**Comment by crawford** [comment]
@danehans Well, the good news is that those useless errors (that's my fault) will soon be replaced with really verbose messages (https://github.com/coreos/fuze/pull/19). A slight correction in terminology, the config you reference is a Fuze Config. bootcfg will parse that and generate an Ignition config which is what is actually passed to the machine. We should probably clean up the error messages in bootcfg, since those terms are technically incorrect.

The error you are getting indicates that there is a path that is relative when it should be absolute (and since my error logging is lacking, we don't know which line it is). Can you double check the Fuze config and make sure that there are no relative paths?

**Comment by danehans** [comment]
@crawford Thanks for the feedback. For troubleshooting purposes I changed from the bootcfg k8s profile and fuze template to the simple/ssh examples and found the root cause of my problem. I have been using a cached Docker image for my bootcfg service that is 3 months old. I removed the cached image and restarted bootcfg, which pulled the latest bootcfg Docker image. The bootcfg example files are now properly working in my lab. Thanks for your help.
